File: talos-firefox/src/suggestions.js
/**
 * Suggestion engine — context-aware ranking, slash hints, and chains.
 *
 * Pure functions extracted from InputBar for testability.
 */

export const BASE_SUGGESTIONS = [
  'Summarize this page',
  'Explain this code',
  'What does this do?',
  'Write a function that ',
  'Help me understand ',
  'How do I ',
  'Fix this bug',
  'Refactor this to ',
  'What are the key points?',
  'Compare these approaches',
  'Give me an example of ',
  'Translate this to ',
  'Debug this error',
  'Optimize this code',
  'What is the difference between ',
  '/search ',
  '/web ',
  '/reason ',
  '/new',
  '/help',
];

export const SLASH_COMMANDS = {
  '/search ': { hint: '<query>', description: 'Search DuckDuckGo' },
  '/web ':    { hint: '<url>',   description: 'Fetch a web page' },
  '/reason ': { hint: '<query>', description: 'Step-by-step reasoning' },
  '/new':     { hint: '',        description: 'New conversation' },
  '/help':    { hint: '',        description: 'Show commands' },
};

/**
 * Detect patterns in assistant response content.
 * @param {string} content — assistant message text
 * @returns {{ hasCode: boolean, hasList: boolean, hasError: boolean, hasExplanation: boolean }}
 */
export function detectResponsePatterns(content) {
  if (!content) return { hasCode: false, hasList: false, hasError: false, hasExplanation: false };
  const c = content.toLowerCase();
  return {
    hasCode: c.includes('```') || c.includes('function') || c.includes('class '),
    hasList: c.includes('1.') || c.includes('- ') || c.includes('* '),
    hasError: c.includes('error') || c.includes('bug') || c.includes('fix') || c.includes('issue'),
    hasExplanation: c.includes('means') || c.includes('because') || c.includes('essentially'),
  };
}

/**
 * Build context-aware follow-up suggestions from last assistant content.
 * @param {string} lastAssistantContent
 * @param {{ url?: string, selection?: string } | null} pageContext
 * @returns {string[]}
 */
export function getContextSuggestions(lastAssistantContent, pageContext = null) {
  if (!lastAssistantContent) return [];

  const { hasCode, hasList, hasError, hasExplanation } = detectResponsePatterns(lastAssistantContent);
  const follow = [];

  if (hasCode) {
    follow.push(
      'Explain this step by step',
      'Can you add error handling?',
      'Write tests for this',
      'Can you optimize this?',
      'Show me a different approach',
    );
  }
  if (hasList) {
    follow.push(
      'Tell me more about the first one',
      'Which do you recommend?',
      'Can you elaborate on that?',
      'Give me a comparison table',
    );
  }
  if (hasError) {
    follow.push(
      'What caused this error?',
      'How do I prevent this?',
      'Are there other edge cases?',
      'Show me the fix',
    );
  }
  if (hasExplanation) {
    follow.push(
      'Can you give me an example?',
      'Explain it more simply',
      'How does this relate to ',
      'What are the tradeoffs?',
    );
  }

  // Page-context-aware suggestions
  if (pageContext) {
    if (pageContext.url && pageContext.url.includes('github.com')) {
      follow.push('What does this PR do?');
    }
    if (pageContext.selection) {
      follow.push('Explain this selection');
    }
  }

  // General follow-ups always available after a response
  follow.push('Go deeper on that', 'Can you rewrite that?', 'Thanks, now ', 'What about ');

  return follow;
}

/**
 * Get slash command parameter hint for current input.
 * @param {string} input
 * @returns {string} — hint text to show after input, or ''
 */
export function getSlashHint(input) {
  if (!input || !input.startsWith('/')) return '';

  // Exact slash command match — show hint
  for (const [cmd, { hint }] of Object.entries(SLASH_COMMANDS)) {
    if (input === cmd) return hint;
  }

  // Partial slash match — complete command + hint
  const lower = input.toLowerCase();
  for (const [cmd, { hint }] of Object.entries(SLASH_COMMANDS)) {
    if (cmd.startsWith(lower) && cmd !== lower) {
      return cmd.slice(input.length) + hint;
    }
  }

  return '';
}

/**
 * Ranked suggestion matching — context suggestions take priority.
 * @param {string} input
 * @param {string[]} contextSuggestions
 * @param {string[]} baseSuggestions
 * @returns {string} — remaining text of best match, or ''
 */
export function rankSuggestions(input, contextSuggestions, baseSuggestions) {
  if (!input || input.length < 2) return '';

  const lower = input.toLowerCase();

  // Slash input — only match slash commands
  if (input.startsWith('/')) {
    const hint = getSlashHint(input);
    if (hint) return hint;
    return '';
  }

  // Context suggestions first
  for (const s of contextSuggestions) {
    if (s.toLowerCase().startsWith(lower)) {
      return s.slice(input.length);
    }
  }

  // Then base suggestions
  for (const s of baseSuggestions) {
    if (s.toLowerCase().startsWith(lower)) {
      return s.slice(input.length);
    }
  }

  return '';
}

// Static follow-up chains: maps a context suggestion to its next suggestion
const CHAIN_MAP = {
  'Explain this step by step': 'Can you simplify that?',
  'Write tests for this': 'Can you add edge cases?',
  'What caused this error?': 'How do I prevent this in the future?',
  'Show me the fix': 'Are there other places this could happen?',
  'Which do you recommend?': 'What are the tradeoffs?',
  'Can you give me an example?': 'Show me a more advanced example',
  'Go deeper on that': 'Can you summarize the key takeaway?',
  'Can you rewrite that?': 'Explain the changes you made',
};

/**
 * Build a suggestion chain — follow-ups to follow-ups.
 * @param {string} lastContent — last assistant content
 * @param {number} depth — 0 = getContextSuggestions, 1 = chain follow-up
 * @returns {string} — preemptive suggestion, or ''
 */
export function buildSuggestionChain(lastContent, depth = 0) {
  if (!lastContent) return '';

  if (depth === 0) {
    const suggestions = getContextSuggestions(lastContent);
    return suggestions.length > 0 ? suggestions[0] : '';
  }

  if (depth === 1) {
    // Look at what the depth-0 suggestion was and find its chain follow-up
    const suggestions = getContextSuggestions(lastContent);
    if (suggestions.length === 0) return '';
    const primary = suggestions[0];
    return CHAIN_MAP[primary] || '';
  }

  return '';
}

File: talos/suggestions.py
"""Ghost suggestion engine for the TUI — mirrors the Firefox extension logic."""

from __future__ import annotations

import re

BASE_SUGGESTIONS: list[str] = [
    "Summarize this page",
    "Explain this code",
    "What does this do?",
    "Write a function that ",
    "Help me understand ",
    "How do I ",
    "Fix this bug",
    "Refactor this to ",
    "What are the key points?",
    "Compare these approaches",
    "Give me an example of ",
    "Debug this error",
    "Optimize this code",
    "What is the difference between ",
]


def detect_response_patterns(content: str) -> dict[str, bool]:
    """Detect patterns in assistant response content."""
    if not content:
        return {"has_code": False, "has_list": False, "has_error": False, "has_explanation": False}
    c = content.lower()
    return {
        "has_code": "```" in c or "function" in c or "class " in c,
        "has_list": "1." in c or "- " in c or "* " in c,
        "has_error": "error" in c or "bug" in c or "fix" in c or "issue" in c,
        "has_explanation": "means" in c or "because" in c or "essentially" in c,
    }


def get_context_suggestions(last_content: str) -> list[str]:
    """Build context-aware follow-up suggestions from last assistant content."""
    if not last_content:
        return []

    patterns = detect_response_patterns(last_content)
    follow: list[str] = []

    if patterns["has_code"]:
        follow.extend([
            "Explain this step by step",
            "Can you add error handling?",
            "Write tests for this",
            "Can you optimize this?",
            "Show me a different approach",
        ])
    if patterns["has_list"]:
        follow.extend([
            "Tell me more about the first one",
            "Which do you recommend?",
            "Can you elaborate on that?",
            "Give me a comparison table",
        ])
    if patterns["has_error"]:
        follow.extend([
            "What caused this error?",
            "How do I prevent this?",
            "Are there other edge cases?",
            "Show me the fix",
        ])
    if patterns["has_explanation"]:
        follow.extend([
            "Can you give me an example?",
            "Explain it more simply",
            "How does this relate to ",
            "What are the tradeoffs?",
        ])

    # General follow-ups always available
    follow.extend(["Go deeper on that", "Can you rewrite that?", "Thanks, now ", "What about "])

    return follow


def match_suggestion(input_text: str, suggestions: list[str]) -> str:
    """Match user input against suggestions (case-insensitive prefix).

    Returns the remaining text of the best match, or empty string.
    """
    if not input_text or len(input_text) < 2:
        return ""
    lower = input_text.lower()
    for s in suggestions:
        if s.lower().startswith(lower):
            return s[len(input_text):]
    return ""


def get_ghost(input_text: str, last_assistant: str = "") -> str:
    """Get the ghost suggestion for the current input.

    - Empty input with context → preemptive (first context suggestion)
    - Typing → prefix-match with context suggestions ranked first, then base
    """
    context_suggestions = get_context_suggestions(last_assistant)

    # Empty input — show preemptive
    if not input_text:
        return context_suggestions[0] if context_suggestions else ""

    # Merge: context first, then base
    all_suggestions = context_suggestions + BASE_SUGGESTIONS
    return match_suggestion(input_text, all_suggestions)
